Fix for Cosmos-SDK 0.44

ctrl-Felix opened the pull request:
After providing a quick and dirty fix last time I now found some time to take a proper approach to this and this should finally fix the compability. This pull request includes:

-  Migrating the transactions to protobuf (Compiled through the [protobuf compiler](https://github.com/protocolbuffers/protobuf/releases/tag/v3.6.1))
- Updating the tests
- Adding a [ibc_transfer method ](https://github.com/ctrl-Felix/cosmospy/blob/master/src/cosmospy/_transaction.py#L73)to allow also ibc transfers
- Add the possibility to add [another account which pays the transaction fee ](https://github.com/ctrl-Felix/cosmospy/blob/e349f80b6bb4d9f7b2ca5ccb2d4850b7440267e0/src/cosmospy/_transaction.py#L38)through the feegrant module
-  Allow transaction broadcasting through the RPC & API as both endpoint need a slightly different payload structure

I think this should be it. All in all the biggest gain is that cosmospy will be working again.

ctrl-Felix committed:
cosmos v0.44 support

add cosmos v0.44 support

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ctrl-Felix committed:
add dependencies

ctrl-Felix committed:
Merge remote-tracking branch 'origin/sdkv0.44' into sdkv0.44

ctrl-Felix committed:
fix hrp for transfers & Update README

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

czarcas7ic committed:
Merge pull request #1 from czarcas7ic/czar

Czar

czarcas7ic committed:
Update _transaction.py

Updates for osmosis testnet faucet

ctrl-Felix committed:
implement feegrant & fix tests

ctrl-Felix committed:
ibc transfers & improved broadcasting

hukkin commented on pyproject.toml:
```diff
@@ -27,6 +27,10 @@ dependencies = [
     "mnemonic >=0.19",
     "hdwallets >=0.1.0",
     "typing-extensions >=3.7.4; python_version < '3.8'",
+    "requests==2.26.0",
+    "protobuf >= 3.19.1",
+    "google >= 3.0.0",
```
```suggestion
    "protobuf >= 3.19.1",
```

As noted in the review of the previous PR, we shouldn't need `requests` or `google` deps.

ctrl-Felix replied:
changed with the two recent commits. Removing requests also involves removing the two helper functions ``broadcast_rpc_tx`` and ``braodcast_api_tx``

hukkin reviewed (changes requested):
Thanks a lot! :partying_face: 

I'll do a proper review in the upcoming days. This is just a drive-by note about the extra dependencies.

Also, if you'd be kind enough to provide the commands/actions that you executed to autogenerate the protobuf files that'd be great!

ctrl-Felix committed:
remove request dependency

ctrl-Felix committed:
remove requests

ctrl-Felix commented:
Thank you very much for your quick reply  😄
I used the protobuf compiler from [here](https://github.com/protocolbuffers/protobuf/releases/tag/v3.6.1 with following command:
Execute in the proto folder
```protoc --python-out=<absolute path to interfaces dir> file.proto ```

Regarding the requests. I added two commits which removes the code snipped which used the requests dependency. (The second one is because I forgot to remove the actual dependency...) 

Adding requests could be useful for helper functions tho. Like a function to get the account sequence and account number which are only accessible through a api call.

ctrl-Felix committed:
fix sync mode assignment

the correct methods for the rpc are: broadcast_tx_sync, broadcast_tx_async and broadcast_tx_commit. The ones for the api are slightly different

ctrl-Felix committed:
fix default sync mode

The default sync mode wasn't correct as it was a default api sync mode but it's better to use rpc sync modes.

ctrl-Felix committed:
update protoc version

ctrl-Felix commented:
Sorry for these commits. I just saw two things while going through the code once again:

- The first one was that I mesed up the broadcast modes for the api and rpc call. While the rpc refers to it as method it wants ``bradcast_tx_sync`` as method. The api refers to it as mode and therefore wants it as ``BROADCAST_MODE_SYNC``
- Second one was that I was using protoc 3.6.1 which worked. But it's 4 years old. I took the recent version now to compile the files and the ward tests go through. However there seeom to be a coverage issue or whatever in the automated github checks. I am not yet sure what that means

hukkin commented on src/cosmospy/_typing.py:
```diff
@@ -1,14 +1,15 @@
 import sys
 
+print(sys.version_info)
```
```suggestion
```

hukkin commented on src/cosmospy/_typing.py:
```diff
@@ -1,14 +1,15 @@
 import sys
```
Why was the `typing` module made private? Can we revert that change?

ctrl-Felix replied:
I took this as it was before without really thinking about it. I will change it

hukkin commented on src/cosmospy/_transaction.py:
```diff
@@ -21,87 +26,149 @@ class Transaction:
+        signer_infos.mode_info.single.mode = 1
+        return signer_infos
+
+    def get_api_pushable(self):
+        api_sync_modes = {
+            "broadcast_tx_sync": "BROADCAST_MODE_SYNC",
+            "broadcast_tx_async": "BROADCAST_MODE_ASYNC",
+            "broadcast_tx_commit": "BROADCAST_MODE_BLOCK"
         }
+        return json.dumps(
+            {"tx_bytes": self._get_tx_bytes(), "mode": api_sync_modes[self._sync_mode]}
+        )
+
+    def get_rpc_pushable(self):
+
+        return json.dumps(
+            {
+                "jsonrpc": "2.0",
+                "id": 1,
+                "method": self._sync_mode,
+                "params": {
+                    "tx": self._get_tx_bytes()
+                }
+            }
+        )
```
```suggestion
```

Let's do as I suggested [here](https://github.com/hukkin/cosmospy/issues/24#issuecomment-1005589846) i.e. remove these functions and instead make `get_tx_bytes` public. Then document in the README how to broadcast via either the Cosmos REST API or Tendermint RPC API.

ctrl-Felix replied:
This would also require some more changes as I think this would make the whole sync mode in typing obsolete and also the param which will be set on the transaction creation

hukkin replied:
Yeah let's remove all of that.

ctrl-Felix replied:
I am currently on it. Also to fix the tests then to use the tx_bytes function then

ctrl-Felix replied:
Actually. You had a get_pushable method. I totally agree not to use requests or any broadcast method in it but I have to admit that the get_pushable method is handy to use

hukkin replied:
It can be handy but not strictly necessary. Given that cosmospy is a "tool for _offline_ transaction signing" I think it was always a bit ugly for it to contain information about the broadcast sync modes. Now that we have more than one broadcast APIs I dislike this bloat even more.

I know things like this may be convenient to hobbyists but enterprise users want to user their own networking libraries and configurations. Hence it makes sense to have a low-level offline library. It may make sense to have another library that is as easy to use as possible, but I have no willingness to maintain that.

ctrl-Felix replied:
Thank you very much for this explanation. I really appreciate it!

hukkin commented on README.md:
```diff
@@ -126,18 +126,20 @@ tx.add_transfer(
-# Optionally submit the transaction using your preferred method.
-# This example uses the httpx library.
-import httpx
+import requests
```
I prefer to keep endorsing the `httpx` library rather than `requests`.

ctrl-Felix replied:
What is the advantage to use httpx? I would say requests is by far the most common library for http requests

hukkin replied:
`httpx` does async (as well as sync) calls.

And yeah I'm free to endorse whichever library I want, am I not? :wink:

ctrl-Felix replied:
I never said something else, lol. That makes sense. I just noticed that while 
```python
rpc_url = "https://rpc.cosmos.network/"
pushable_tx = json.dumps({
                "jsonrpc": "2.0",
                "id": 1,
                "method": "broadcast_tx_sync", # Available methods: broadcast_tx_sync, broadcast_tx_async, broadcast_tx_commit
                "params": {
                    "tx": tx_bytes
                }
              })

r = requests.post(rpc_url, data=pushable_tx)
``` 
works httpx wasn't:
```python
rpc_url = "https://rpc.cosmos.network/"
pushable_tx = {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "broadcast_tx_sync", # Available methods: broadcast_tx_sync, broadcast_tx_async, broadcast_tx_commit
                "params": {
                    "tx": tx_bytes
                }
              }

r = httpx.post(rpc_url, data=pushable_tx)
```

Which would give me:
```
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32700,
    "message": "Parse error. Invalid JSON",
    "data": "error unmarshaling request: invalid character 'j' looking for beginning of value"
  }
}
```

Well, seems like I have to see what's wrong

hukkin replied:
The httpx example is missing a `json.dumps` isn't it?

ctrl-Felix replied:
You're absolutely right.. .And that's how I initially made it until Pycharm complained that it got a str instead of a dict. Well, it works now

hukkin commented on README.md:
```diff
@@ -126,18 +126,20 @@ tx.add_transfer(
 
-api_base_url = "https://node.atomscan.com"
-httpx.post(api_base_url + "/txs", data=pushable_tx)
+# Submit the transaction through the rpc
```
```suggestion
# Submit the transaction through the Tendermint RPC
```

hukkin commented on README.md:
```diff
@@ -126,18 +126,20 @@ tx.add_transfer(
+pushable_tx = tx.get_rpc_pushable()
+r = requests.post(rpc_url, data=pushable_tx)
+
+# Submit the transaction through the api
```
```suggestion
# Submit the transaction through the Cosmos REST API
```

hukkin commented on README.md:
```diff
@@ -126,18 +126,20 @@ tx.add_transfer(
+# Submit the transaction through the api
+rpc_api = "https://api.cosmos.network/"
+pushable_tx = tx.get_api_pushable()
+r = requests.post(rpc_api, data=pushable_tx)
```
Isn't the resource missing from the URI? I.e. `/cosmos/tx/v1beta1/txs`.

hukkin commented on src/cosmospy/_transaction.py:
```diff
@@ -21,87 +26,149 @@ class Transaction:
+        msg_any.type_url = "/cosmos.bank.v1beta1.MsgSend"
+        self._tx_body.messages.append(msg_any)
+
+    def add_ibc_transfer(
```
This seems like new API/functionality unrelated to the deprecated endpoint. Can we design, add and test it in a separate PR?

ctrl-Felix replied:
Yes, we can do that too.  We then need to also change the tests because I already added a test for this tx type.

hukkin commented on src/cosmospy/_transaction.py:
```diff
@@ -21,87 +26,149 @@ class Transaction:
+            fee: int,
+            gas: int,
+            fee_denom: str = "uatom",
+            fee_account: str = "",
```
`fee_account` wasn't needed before. Do we need it now? Is it new functionality that can be added in another PR?

ctrl-Felix replied:
It was indeed a new functionality implementing the feegrant module. I implemented this because I needed it for a project and it is just one attribute that has to be passed to the transaction. We can make a separate PR for this with a better feegrant implementation then.

ctrl-Felix committed:
Update src/cosmospy/_typing.py

Co-authored-by: Taneli Hukkinen <3275109+hukkin@users.noreply.github.com>

ctrl-Felix committed:
Update README.md

Co-authored-by: Taneli Hukkinen <3275109+hukkin@users.noreply.github.com>

ctrl-Felix committed:
Update README.md

Co-authored-by: Taneli Hukkinen <3275109+hukkin@users.noreply.github.com>

ctrl-Felix committed:
Update src/cosmospy/_transaction.py

Co-authored-by: Taneli Hukkinen <3275109+hukkin@users.noreply.github.com>

ctrl-Felix committed:
changes as discussed in #32

ctrl-Felix commented:
I went through everything now I think. I have to admit I am not 100% happy by removing the get_rpc_pushable and get_api_pushable methods, especially as this method was implemented before too.

hukkin commented on src/cosmospy/_transaction.py:
```diff
@@ -21,87 +24,96 @@ class Transaction:
+            self, recipient: str, amount: int, denom: str = "uatom", hrp: str = DEFAULT_BECH32_HRP
+    ) -> None:
+        msg = transfer.MsgSend()
+        msg.from_address = privkey_to_address(self._privkey, hrp=hrp)
```
```suggestion
        msg.from_address = privkey_to_address(self._privkey, hrp=self._hrp)
```

hukkin commented on src/cosmospy/_transaction.py:
```diff
@@ -21,87 +24,96 @@ class Transaction:
+        self._tx_raw = tx.TxRaw()
 
+    def add_transfer(
+            self, recipient: str, amount: int, denom: str = "uatom", hrp: str = DEFAULT_BECH32_HRP
```
```suggestion
            self, recipient: str, amount: int, denom: str = "uatom"
```

hukkin reviewed (changes requested):
Let's revert this unless there's a good reason for the change?

ctrl-Felix committed:
Update src/cosmospy/_transaction.py

Co-authored-by: Taneli Hukkinen <3275109+hukkin@users.noreply.github.com>

ctrl-Felix committed:
Update src/cosmospy/_transaction.py

Co-authored-by: Taneli Hukkinen <3275109+hukkin@users.noreply.github.com>

ctrl-Felix committed:
remove unused dependencies

ctrl-Felix committed:
Merge pull request #1 from ctrl-Felix/ctrl-felix/patch-2

remove unused dependencies

ctrl-Felix committed:
remove ibc deps

hukkin reviewed (changes requested):
Could you sync the proto files (and generate .py files) with what's up-to-date in https://github.com/cosmos/cosmos-sdk/tree/main/proto ?

hukkin commented on src/cosmospy/_transaction.py:
```diff
@@ -2,13 +2,15 @@
 
 from cosmospy._wallet import DEFAULT_BECH32_HRP, privkey_to_address, privkey_to_pubkey
-from cosmospy.typing import SyncMode
+import cosmospy.interfaces.any_pb2 as Any
```
```suggestion
import cosmospy._interfaces.any_pb2 as Any
```

Can we make the `interfaces` package private?

ctrl-Felix committed:
recompile proto files & make interfaces private

ctrl-Felix commented:
Here are the sources for the new proto files: 
* https://github.com/cosmos/gogoproto/blob/master/gogoproto/gogo.proto
* https://github.com/cosmos/cosmos-sdk/blob/main/proto/cosmos/msg/v1/msg.proto
* https://github.com/cosmos/cosmos-proto/blob/main/proto/cosmos_proto/cosmos.proto

I removed the functions for MultiSig as well as other transaction types and only took MsgSend to keep the dependencies as low as possible

Changes I made to every .proto file:
* Fix the imports to the dir structure in cosmospy

Changes I made to every compiled proto file:
* fix the imports as discussed in https://github.com/protocolbuffers/protobuf/issues/1491 (protoc isn't compiling them correctly)

ctrl-Felix commented:
Hi @hukkin 
I just wanted to ask if you had time to check the new commits
